services/web/server/src/simcore_service_webserver/folders/_workspaces_api.py

```diff
@@ -0,0 +1,138 @@
+import logging
+
+from aiohttp import web
+from models_library.folders import FolderID
+from models_library.products import ProductName
+from models_library.users import UserID
+from models_library.workspaces import WorkspaceID
+from simcore_postgres_database.utils_repos import transaction_context
+
+from ..db.plugin import get_asyncpg_engine
+from ..projects import _db_v2 as projects_db
+from ..projects import _folders_db as project_to_folders_db
+from ..projects import _groups_db as project_groups_db
+from ..projects._access_rights_api import check_user_project_permission
+from ..users.api import get_user
+from ..workspaces.api import check_user_workspace_access
+from . import _folders_db
+
+_logger = logging.getLogger(__name__)
+
+
+async def move_folder_into_workspace(
+    app: web.Application,
+    *,
+    user_id: UserID,
+    folder_id: FolderID,
+    workspace_id: WorkspaceID | None,
+    product_name: ProductName,
+) -> None:
+    # 1. User needs to have delete permission on source folder
+    folder_db = await _folders_db.get(
+        app, folder_id=folder_id, product_name=product_name
+    )
+    workspace_is_private = True
+    if folder_db.workspace_id:
+        await check_user_workspace_access(
+            app,
+            user_id=user_id,
+            workspace_id=folder_db.workspace_id,
+            product_name=product_name,
+            permission="delete",
+        )
+        workspace_is_private = False
+
+    # 2. User needs to have write permission on destination workspace
+    if workspace_id is not None:
+        await check_user_workspace_access(
+            app,
+            user_id=user_id,
+            workspace_id=workspace_id,
+            product_name=product_name,
+            permission="write",
+        )
+
+    # 3. User needs to have delete permission on all the projects inside source folder
+    (
+        folder_ids,
+        project_ids,
+    ) = await _folders_db.get_all_folders_and_projects_recursively(
+        app,
+        connection=None,
+        folder_id=folder_id,
+        private_workspace_user_id_or_none=user_id if workspace_is_private else None,
+        product_name=product_name,
+    )
+    # NOTE: Not the most effective, can be improved
+    for project_id in project_ids:
+        await check_user_project_permission(
+            app,
+            project_id=project_id,
+            user_id=user_id,
+            product_name=product_name,
+            permission="delete",
+        )
+
+    # ⬆️ Here we have already guaranties that user has all the right permissions to do this operation ⬆️
+
+    async with transaction_context(get_asyncpg_engine(app)) as conn:
+        # 4. Update workspace ID on the project resource
+        for project_id in project_ids:
+            await projects_db.patch_project(
```

Comment:
again this entry is asking for a batch alternative ... the code is talking to us ;-)